PR dead: restic backup: limit maximum size of repository during backup - new implementation with PR #5338

wplapper opened the pull request:
A new option --max-repo-size is introduced and then passed on to repository.saveAndEncrypt()
The current size of the repository is monitored when a blob gets saved. If the maximum is exceeded, a fatal error will be triggered.

The snapshot gets terminated gracefully and a new partial snapshot will be added to the repository.

<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------
The PR adresses https://github.com/restic/restic/issues/4585

<!--
Describe the changes and their purpose here, as detailed as needed.
-->

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
https://github.com/restic/restic/issues/4585
<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->

Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].

Please always follow these steps:
- Read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- Enable [maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- Run `gofmt` on the code in all commits.
- Format all commit messages in the same style as [the other commits in the repository](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
-->

- [x] I have added tests for some code changes - more to come!
- [ ] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [ ] I'm done! This pull request is ready for review.

MichaelEischer commented on changelog/unreleased/issue-4585:
```diff
@@ -0,0 +1,17 @@
+linked to a new snapshot. Therefore they are unreferenced, but
+are still occupying space in tge repository.
+
+I need help with shutting down the backup process properly, so the failed backup
```
What's that doing in the changelog entry?

MichaelEischer replied:
Restic currently doesn't have the concept of a partial snapshot. Storing such a snapshot will currently cause the next backup run to re-read all files that didn't make it into the partial snapshot.

Adding a corresponding flag to the snapshot shouldn't be too hard. Similarly, the archiver and fileSaver could be adapted to stop reading new files. Then the whole backup process would quickly unwind itself and finish with a partial snapshot. (This should happen in a separate PR as partial snapshots are a useful feature in itself). Afterwards, we can think about repository size limits again. The main remaining complexity would be estimating how much space would still be necessary to write the remaining index and tree blobs when approaching the size limit.

MichaelEischer commented on internal/archiver/archiver.go:
```diff
@@ -508,6 +513,13 @@ func (arch *Archiver) save(ctx context.Context, snPath, target string, previous
 	case fi.Mode.IsRegular():
 		debug.Log("  %v regular file", target)
 
+		// check if repository size is about to be exceeded.
```
This can only check after the fact, but not beforehand. Reading from the summary also requires taking the `arch.mu` mutex.

But there's a larger problem here. When someone backs up very large files, this can miss the target size by hundreds of gigabytes (assuming such large files).

MichaelEischer reviewed (commented):
I've added a few comments regarding the approach. In short, a proper implementation will require a significant amount of supporting infrastructure.

wplapper commented:
I put this PR on hold, since I do not understand the full complexity of all the issues involved.

wplapper commented:
Working on a completely new implementation.

wplapper committed:
new implementation of limiting repository size during backup

Changed saveAndEncrypt(...) to include tracking of current size and
raising an error "MaxCapacityExceeded" when the limit has been exceeded.

Added CurrentRepositorySize(ctx) to report current size

Added MaxCapacityExceeded() to query if limit has been execced

Added IsRepositoryLimitActive() to query if size monitoring is active

In addition an interface definition added to fulfill the needs of internal/archiver
for accessing repository functions.

The issue descrition has been updated.

wplapper committed:
In preparation for `restic backup --max-repo-size` extend interface definition

of repository to include:
  MaxCapacityExceeded() bool
  IsRepositoryLimitActive() bool

wplapper committed:
Add field RepoSizeMax to global.go, Add field PartialSnapshot to Snapshot

in preparation to limit repository size during backup.

wplapper committed:
restic backup - handle MaxCapacityExceeded while saving a blob

If the capacity size limit condition is activated, a TreeBlob is passed through
unharmed, so the new tree can be saved properly.

A DataBlob however, which could be very large in size will be replaced by the
data blob which contains the string "MaxCapacityExceeded\n".

wplapper committed:
changed logic for processing failing MaxCapacityExceeded data blobs

intercept MaxCapacityExceeded errors for all blobs and convert them to "no error" for
all blobs but DataBlobs.

For DataBlobs insert an extra data blob once, with the signature "MaxCapacityExceeded\n"

For all data blob instances with the above error, return a `saveBlobResponse` with an ID of
restic.Hash("MaxCapacityExceeded\n")

This done to maintain the integritry of all newly saved blobs.

wplapper committed:
changes to internal/archiver/archiver: allow `restic backup` to work with a repo size limitation

in saveDir() don't save the directory if the current backup is in repository shutdown mode

In saveTree() don't descend into subdirectories if the current backup is in repository shutdown mode

in Snapshot() add the flag PartialSnapshot to the Snapshot structure, also add tag "partial-snapshot"
to the tag list.

wplapper committed:
restic backup --max-repo-size

Implementation of limiting the repository size during the backup process.

Introduce option --max-repo-size, which is converted to field
gopts.RepoSizeMax, and handed over to repository.New() in cmd/restic/global.go.

The command prints the current size in Verbosef mode at the beginning and
also at the end, when the repository size has been exceeded.

wplapper committed:
restic backup - remove unneeded code and improve error information

remove unneded field RepoSizeMax

Improve code at the end of the run showing if repository capacity has been
exceeded and give current value.

wplapper committed:
internal/archiver/blob_saver removed ugly hack to save data blobs

in case of the error "MaxCapacityExceeded", now just ignore it and continue
as before.

wplapper committed:
repository functions:

repo/repository.go: print limit reached message once, remove unneedef function IsRepositoryLimitActive()
restic/repository.go: remove unnneded interface definition

wplapper committed:
filesaver in repository size monitoring mode

If in monitoring mode, an active file gets truncated by breaking out of the reead loop
The incomplete file list gets fixed by truncating it to the last non-null ID

The file_saver_test.go needs updating with the repo parameter.

wplapper committed:
archiver in repository size monitoring mode

archiver.go: add repo parameter to newFileSaver(), remove unneeded check for monitoring check for saveTree

blob_saver.go: remove unneded monitoring check for worker()

wplapper committed:
restic backup with repository size monitoring: enlarged help description

added function checkPartialSnapshot() so commands working on snapshots
can check ig they can use a given partial snapshot as input.

wplapper committed:
restic backup - gofmt got me

here is the better version by gofmt

wplapper committed:
the commands which will definitely not operate on partial snapshots

wplapper committed:
restic diff, stats, tag - more command which vannot handle partial snapshots

inserted check for partial snapshots

wplapper committed:
restic prune - added function findPartialSnapshots()

findPartialSnapshots collects all partial snapshots and sends the off to
restic.ParallelRemove() to forget them before prune starts working.

wplapper committed:
restic prune integration tests:

added one test TestPruneSizeMonitoring() which runs a failed backup and then calls prune
to remove the partial snapshot.

cmd/restic/integration_test needed to be modified, since it does not allow multiple calls
to list snapshots.

wplapper committed:
repository/repository: fixed race condition i

by having a global lock attached to the Repository structure

wplapper committed:
repository/repository - fix unused sync.Mutex m

wplapper committed:
repository/repository - protect r.maxRepoCapReached

by Lock() in MaxCapacityExceeded()

wplapper committed:
restic backup - corrected typo in error message

wplapper committed:
restic backup and restic prune: more / modified tests

cmd_prune_integration_test: testing for the exact error message

cmd_backup_integration_test: two additional backup test with repository size monitoring
TestBackupRepoSizeMonitorOnFirstBackup(): limit first backup on size,
check error message

TestBackupRepoSizeMonitorOnSecondBackup(): do a normal first backup and then limit second
backup, so size is already exceeded on start of backup: check for error message.

wplapper commented:
I will close this PR, since it contains too many changes.I will split it as multiple PRs, starting off with a partial snapshot.

wplapper committed:
updated logic to limit repository size
